### PR TITLE
Implement create notification hub method

### PR DESCRIPTION
**Pull request by varvusc**

### CODE REVIEW

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 1]
```diff
@@ -183,32 +193,84 @@ public NamespaceManager(IEnumerable<Uri> addresses, NamespaceManagerSettings set
...
-        public Task<NotificationHubDescription> CreateNotificationHubAsync(NotificationHubDescription description)
+        public async Task<NotificationHubDescription> CreateNotificationHubAsync(NotificationHubDescription description)
         {
-            throw new NotImplementedException();
+            if (description == null)
+            {
+                throw new ArgumentNullException(nameof(description));
+            }
+
+            var client = new HttpClient();
```
Comment (iTOYS) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Use `using`

```suggestion
            using(var client = new HttpClient())
            {}
```

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 2]
```diff
@@ -183,32 +193,84 @@ public NamespaceManager(IEnumerable<Uri> addresses, NamespaceManagerSettings set
...
+                    httpRequestMessage.Content = new StreamContent(new MemoryStream(Encoding.UTF8.GetBytes(xmlBody)));
+                    httpRequestMessage.Headers.Add("Authorization", token);
+                    httpRequestMessage.Headers.Add("x-ms-version", ApiVersion);
+
+                    return await client.SendAsync(httpRequestMessage);
+                });
+
+            if (response.IsSuccessStatusCode)
+            {
+                var xmlResponse = await GetXmlContent(response);
```
Comment (iTOYS) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Dispose?

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 3]
```diff
@@ -183,32 +193,84 @@ public NamespaceManager(IEnumerable<Uri> addresses, NamespaceManagerSettings set
...
+                return GetModelFromResponse<NotificationHubDescription>(xmlResponse);
+            }
+            else
+            {
+                var xmlError = await GetXmlError(response);
```
Comment (iTOYS) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Dispose?

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 4]
```diff
@@ -183,32 +193,84 @@ public NamespaceManager(IEnumerable<Uri> addresses, NamespaceManagerSettings set
...
+                var error = GetModelFromResponse<ErrorResponse>(xmlError);
+
+                var innerException = new WebException($"The remote server returned an error: {error.Code}");
+                throw new MessagingEntityAlreadyExistsException(error.Detail, innerException);
```
Comment (iTOYS) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: I believe that we can get not only `409` status code

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 5]
```diff
@@ -273,5 +335,53 @@ public Task<NotificationHubDescription> UpdateNotificationHubAsync(NotificationH
...
+        private async Task<XmlReader> GetXmlError(HttpResponseMessage response)
+        {
+            var reader = XmlReader.Create(await response.Content.ReadAsStreamAsync());
+            return reader;
+        }
+
+        private T GetModelFromResponse<T>(XmlReader xmlReader) where T : class
+        {
+            var serializer = new DataContractSerializer(typeof(T));
+            return (T)serializer.ReadObject(xmlReader);
```
Comment (iTOYS) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]:
```suggestion
            using(xmlReader)
            {
                return (T)serializer.ReadObject(xmlReader);
            }
```

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 6]
```diff
@@ -273,5 +349,57 @@ public Task<NotificationHubDescription> UpdateNotificationHubAsync(NotificationH
...
+
+            using (var xmlWriter = XmlWriter.Create(stringBuilder, new XmlWriterSettings { OmitXmlDeclaration = true }))
+            {
+                serializer.WriteObject(xmlWriter, model);
+            }
+
+            return stringBuilder.ToString();
+        }
+
+        private async Task<XmlReader> GetXmlContent(HttpResponseMessage response)
```
Comment (mpodwysocki) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Can probably be made static as it is not using any member variables

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 7]
```diff
@@ -273,5 +349,57 @@ public Task<NotificationHubDescription> UpdateNotificationHubAsync(NotificationH
...
+                if (xmlReader.ReadToDescendant("content"))
+                {
+                    xmlReader.ReadStartElement();
+                }
+            }
+
+            return xmlReader;
+        }
+
+        private async Task<XmlReader> GetXmlError(HttpResponseMessage response)
```
Comment (mpodwysocki) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Can probably be made static as it is not using any member variables

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 8]
```diff
@@ -273,5 +349,57 @@ public Task<NotificationHubDescription> UpdateNotificationHubAsync(NotificationH
...
+        {
+            var reader = XmlReader.Create(await response.Content.ReadAsStreamAsync());
```
Comment (mpodwysocki) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Can probably be collapsed into a single line

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 9]
```diff
@@ -273,5 +349,57 @@ public Task<NotificationHubDescription> UpdateNotificationHubAsync(NotificationH
...
+            return reader;
+        }
+
+        private T GetModelFromResponse<T>(XmlReader xmlReader) where T : class
```
Comment (mpodwysocki) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Can probably be made static as it is not using any member variables

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 10]
```diff
@@ -273,5 +349,57 @@ public Task<NotificationHubDescription> UpdateNotificationHubAsync(NotificationH
...
+
+        private string AddHeaderAndFooterToXml(string content)
+        {
+            var header = "<?xml version=\"1.0\" encoding=\"utf-8\"?><entry xmlns = \"http://www.w3.org/2005/Atom\"><content type = \"application/xml\">";
+            var footer = "</content></entry>";
+
+            return $"{header}{content}{footer}";
+        }
+
+        private string SerializeObject<T>(T model)
```
Comment (mpodwysocki) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Can probably be made static as it is not using any member variables

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 11]
```diff
@@ -273,5 +349,57 @@ public Task<NotificationHubDescription> UpdateNotificationHubAsync(NotificationH
         {
             throw new NotImplementedException();
         }
+
+        private string AddHeaderAndFooterToXml(string content)
```
Comment (mpodwysocki) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Can probably be made static as it is not using any member variables

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 12]
```diff
@@ -273,5 +349,57 @@ public Task<NotificationHubDescription> UpdateNotificationHubAsync(NotificationH
...
+        {
+            var header = "<?xml version=\"1.0\" encoding=\"utf-8\"?><entry xmlns = \"http://www.w3.org/2005/Atom\"><content type = \"application/xml\">";
```
Comment (mpodwysocki) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: Can convert these into constant strings

File: src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs [hunk 13]
```diff
@@ -183,32 +193,98 @@ public NamespaceManager(IEnumerable<Uri> addresses, NamespaceManagerSettings set
...
-        {
-            throw new NotImplementedException();
-        }
+        /// <summary>
+        /// Creates a notification hub.
+        /// </summary>
+        /// <param name="hubName">The notification hub description name.</param>
+        /// <returns>An instance of the <see cref="NotificationHubDescription"/> class</returns>
+        public async Task<NotificationHubDescription> CreateNotificationHubAsync(string hubName) => 
+            await CreateNotificationHubAsync(new NotificationHubDescription(hubName));
```
Comment (mpodwysocki) [inline src/Microsoft.Azure.NotificationHubs/NamespaceManager.cs]: `async` and `await` should not be necessary here

**Pull request merged**